olp-cpp-sdk-core/tests/olpclient/unittest/OlpClientHttpTest.cpp

```diff
@@ -0,0 +1,1213 @@
+                                      std::multimap<std::string, std::string>(),
+                                      nullptr, std::string(), callback);
+  auto response = p.get_future().get();
+  ASSERT_EQ(olp::http::ErrorCode::INVALID_URL_ERROR,
+            static_cast<olp::http::ErrorCode>(response.status));
+}
+
+TEST_P(ClientHttp, numberOfAttempts) {
+  client_settings_.retry_settings.max_attempts = 6;
+  client_settings_.retry_settings.retry_condition =
+      ([](const olp::client::HttpResponse&) { return true; });
+
+  auto network = std::make_shared<NetworkMock>();
+  client_settings_.network_request_handler = network;
+  EXPECT_CALL(*network, Send(_, _, _, _, _))
+      .Times(6)
+      .WillRepeatedly([&](olp::http::NetworkRequest request,
+                          olp::http::Network::Payload payload,
+                          olp::http::Network::Callback callback,
+                          olp::http::Network::HeaderCallback header_callback,
+                          olp::http::Network::DataCallback data_callback) {
+        callback(olp::http::NetworkResponse().WithStatus(429));
+
+        return olp::http::SendOutcome(olp::http::RequestId(5));
+      });
+
+  client_.SetSettings(client_settings_);
+
+  std::promise<olp::client::HttpResponse> promise;
+  olp::client::NetworkAsyncCallback callback =
+      [&](olp::client::HttpResponse http_response) {
+        promise.set_value(http_response);
+      };
+
+  auto cancel_token = client_.CallApi(std::string(), "GET",
+                                      std::multimap<std::string, std::string>(),
+                                      std::multimap<std::string, std::string>(),
+                                      std::multimap<std::string, std::string>(),
+                                      nullptr, std::string(), callback);
+
+  auto response = promise.get_future().get();
+  ASSERT_EQ(429, response.status);
+}
+
+TEST_P(ClientHttp, ZeroAttempts) {
+  client_settings_.retry_settings.max_attempts = 0;
+  client_settings_.retry_settings.retry_condition =
+      ([](const olp::client::HttpResponse&) { return true; });
+
+  auto network = std::make_shared<NetworkMock>();
+  client_settings_.network_request_handler = network;
+  EXPECT_CALL(*network, Send(_, _, _, _, _))
+      .WillOnce([&](olp::http::NetworkRequest request,
+                    olp::http::Network::Payload payload,
+                    olp::http::Network::Callback callback,
+                    olp::http::Network::HeaderCallback header_callback,
+                    olp::http::Network::DataCallback data_callback) {
+        callback(olp::http::NetworkResponse().WithStatus(429));
+
+        return olp::http::SendOutcome(olp::http::RequestId(5));
+      });
+
+  client_.SetSettings(client_settings_);
+
+  std::promise<olp::client::HttpResponse> promise;
+  olp::client::NetworkAsyncCallback callback =
+      [&](olp::client::HttpResponse http_response) {
```

Comment:
const ref for the response and probably not global capture, but only for promise?

Reply:
Actually, I prefer global captures as it reduces drastically the size of the lambda and makes sure that you don't capture something that is not used. Let the compiler capture what is needed.